presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java

```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
 
-            // replace the tablescan node with a values node
-            ValuesNode valuesNode = new ValuesNode(idAllocator.getNextId(), inputs, rowsBuilder.build());
-            return SimplePlanRewriter.rewriteWith(new Replacer(valuesNode), node);
+            Assignments.Builder assignments = Assignments.builder();
+            for (int i = 0; i < node.getOutputVariables().size(); i++) {
+                assignments.put(node.getOutputVariables().get(i), scalars.get(i));
+            }
+            ValuesNode valuesNode = new ValuesNode(idAllocator.getNextId(), inputs, ImmutableList.of(scalars));
+            return new ProjectNode(idAllocator.getNextId(), valuesNode, assignments.build());
+        }
+
+        private boolean isReducible(AggregationNode node)
+        {
+            if (node.getAggregations().isEmpty() || !(node.getSource() instanceof TableScanNode)) {
+                return false;
+            }
+            for (Aggregation aggregation : node.getAggregations().values()) {
+                FunctionMetadata functionMetadata = metadata.getFunctionManager().getFunctionMetadata(aggregation.getFunctionHandle());
+                if (!AGGREGATION_SCALAR_MAPPING.containsKey(functionMetadata.getName()) || functionMetadata.getArgumentTypes().size() > 1) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private RowExpression reduce(FunctionMetadata aggregationFunctionMetadata, List<RowExpression> arguments)
+        {
+            QualifiedFunctionName scalarFunctionName = AGGREGATION_SCALAR_MAPPING.get(aggregationFunctionMetadata.getName());
+            Type returnType = metadata.getTypeManager().getType(aggregationFunctionMetadata.getReturnType());
+            while (arguments.size() > 1) {
+                List<RowExpression> reducedArguments = new ArrayList<>();
+                // We fold for every 100 values because GREATEST/LEAST has argument count limit
+                for (List<RowExpression> partitionedArguments : Lists.partition(arguments, 100)) {
+                    List<TypeSignatureProvider> typeSignatureProviders = partitionedArguments.stream()
+                            .map(RowExpression::getType)
+                            .map(Type::getTypeSignature)
+                            .map(TypeSignatureProvider::new)
+                            .collect(toImmutableList());
+                    Object reducedValue = evaluateConstantRowExpression(
+                            new CallExpression(
+                                    scalarFunctionName.getFunctionName(),
+                                    metadata.getFunctionManager().resolveFunction(Optional.empty(), scalarFunctionName, typeSignatureProviders),
+                                    returnType,
+                                    partitionedArguments),
+                            metadata,
+                            session.toConnectorSession());
```

Comment:
Use the util in `Expressions` to build `CallExpression`

Reply:
Not aware of this. Looks so much better.